File: 0658-find-k-closest-elements/0658-find-k-closest-elements.py
class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        result = list()

        if arr is None or len(arr) == 0:
            return result
        
        low = 0
        high = len(arr) - k

        while low < high:
            mid = low + (high - low) // 2

            distS = x - arr[mid]
            distE = arr[mid + k] - x
            if distS > distE:
                low = mid + 1
            else:
                high = mid
        
        return arr[low: low + k]
# Binary Search
# TC =O(log(n - k)).
# Space complexity : O(1)

File: 0658-find-k-closest-elements/0658-find-k-closest-elements.java
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> result = new ArrayList<>();
      
        // Binary search to find the closest position to x
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] < x) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        // Initialize two pointers around the closest position
        left = right - 1;
        
        // Expand around the closest position to find k elements
        while (k > 0) {
            if (left < 0) {
                right++;
            } else if (right >= arr.length || Math.abs(arr[left] - x) <= Math.abs(arr[right] - x)) {
                left--;
            } else {
                right++;
            }
            k--;
        }

        // Collect the elements between left and right pointers
        for (int i = left + 1; i < right; i++) {
            result.add(arr[i]);
        }
        
        return result;
    }
}
